File: Simple Linear regression/Qn1.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

#Importing the data :
wg=pd.read_csv('calories_consumed.csv')
wg.head()

#Checking column names:
c=1
for i in wg.columns:
    print(c,'.',i)
    c+=1

#Checking null values : 
wg.info()
wg.isnull().sum()

#Segregating mput and output data:
x=wg.iloc[:,1].values
y=wg.iloc[:,0].values

#Checking Correlation :
wg.corr()
plt.scatter(np.log(wg['Calories Consumed']),wg['Weight gained (grams)'])
plt.scatter(wg['Calories Consumed'],np.log(wg['Weight gained (grams)']))
sns.pairplot(wg)

#EDA : 
wg.describe()

for i in wg.columns:
    print('Mean for {} is {}'.format(i,wg[i].mean()))
    print('Median for {} is {}'.format(i,wg[i].median()))
    print('Mode for {} is {}'.format(i,wg[i].mode()))
    
   #Univariate Analysis for numerical data :
for i in wg.columns:
    print('column:',i)
    print('Skewness for {} is {}'.format(i,wg[i].skew()))
    print('Kurtosis for{} is {}'.format(i,wg[i].kurt()))
    sns.distplot(wg[i])
    plt.show
     
   #Violin Plot :
f,ax=plt.subplots(figsize=(7,5))
sns.violinplot(wg['Calories Consumed'],wg['Weight gained (grams)'])
plt.show()

   #Box Plot :
f,ax=plt.subplots(figsize=(7,5))
sns.boxplot(wg['Calories Consumed'],wg['Weight gained (grams)'])
plt.show()

   #Boxen Plot :
f,ax=plt.subplots(figsize=(7,5))
sns.boxenplot(wg['Calories Consumed'],wg['Weight gained (grams)'])
plt.show()

#Model building 
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=0)

from sklearn.linear_model import LinearRegression
regressor=LinearRegression()
regressor.fit(x_train.reshape(-1,1),y_train.reshape(-1,1))

y_pred=regressor.predict(x_test.reshape(-1,1))

from sklearn.metrics import r2_score
score=r2_score(y_test,y_pred)
from sklearn.metrics import mean_squared_error
from math import sqrt
rmse=sqrt(mean_squared_error(y_test,y_pred))

#log transformation

np.log(wg['Calories Consumed']).skew()
np.log(wg['Weight gained (grams)']).skew()
plt.hist(np.log(wg['Calories Consumed']))
plt.hist(np.log(wg['Weight gained (grams)']))

wg_log=np.log(wg)

xl=wg_log.iloc[:,1].values

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(xl,y,test_size=0.3,random_state=0)

from sklearn.linear_model import LinearRegression
regressor=LinearRegression()
regressor.fit(x_train.reshape(-1,1),y_train.reshape(-1,1))

yl_pred=regressor.predict(x_test.reshape(-1,1))

from sklearn.metrics import r2_score
score_l=r2_score(y_test,yl_pred)
from sklearn.metrics import mean_squared_error
from math import sqrt
rmse_log=sqrt(mean_squared_error(y_test,yl_pred))

#Exponential transformation

yexp=wg_log.iloc[:,0].values

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,yexp,test_size=0.3,random_state=0)

from sklearn.linear_model import LinearRegression
regressor=LinearRegression()
regressor.fit(x_train.reshape(-1,1),y_train.reshape(-1,1))

y_pred=regressor.predict(x_test.reshape(-1,1))

from sklearn.metrics import r2_score
score_ex=r2_score(y_test,y_pred)
from sklearn.metrics import mean_squared_error
from math import sqrt
rmse_exp=sqrt(mean_squared_error(y_test,yl_pred))

File: Simple Linear regression/Qn1.R
library(MASS)
library(faraway)
wgcc=read.csv(file.choose())
attach(wgcc)

#EDA:
hist(wgcc$Weight.gained..grams.)
boxplot(Weight.gained..grams.)
y<-boxplot(Weight.gained..grams.)
y$out
mean(Weight.gained..grams.)
median(Weight.gained..grams.)
hist(wgcc$Calories.Consumed)
boxplot(Calories.Consumed)
z<-boxplot(Calories.Consumed)
z$out
mean(Calories.Consumed)
median(Calories.Consumed)

#Modeling:
colnames(wgcc)
plot(Calories.Consumed,Weight.gained..grams.)
cor(Calories.Consumed,Weight.gained..grams.)
model1=lm(Weight.gained..grams.~Calories.Consumed)
model1
summary(model1)
predict(model1)
model1$residuals
confint(model1,level = 0.95)
predict(model1,interval = 'confidence')
rmse=sqrt(mean(model1$residuals^2))
rmse

#Log Model
plot(log(Calories.Consumed),Weight.gained..grams.)
cor(log(Calories.Consumed),Weight.gained..grams.)
model2=lm(Weight.gained..grams.~log(Calories.Consumed))
model2
summary(model2)
rmse2=sqrt(mean(model2$residuals^2))
rmse2

#Exponetial Model
plot(Calories.Consumed,log(Weight.gained..grams.))
cor(Calories.Consumed,log(Weight.gained..grams.))
model3=lm(log(Weight.gained..grams.)~Calories.Consumed)
summary(model3)
model3$residuals
log_cc=predict(model3,interval = 'confidence')
log_cc
cc=exp(log_cc)
err1=Weight.gained..grams.-cc
err1
rmse3=sqrt(mean(err1^2))
rmse3

#polynomial transformation
model4=lm(log(Weight.gained..grams.)~Calories.Consumed+I(Calories.Consumed*Calories.Consumed))
summary(model4)
confint(model4,level=0.95)    
log_pwg=predict(model4,interval = 'confidence')
log_pwg
wgpoly=exp(log_pwg)
err_poly=Weight.gained..grams.-wgpoly
err_poly
rmse4=sqrt(mean(err_poly^2))
rmse4
rmse3
rmse2
rmse

#Box-cox transformation
model5=boxcox(model1,lambda = seq(-3,3))
plot(model1)
best.lam=model5$x[which(model5$y==max(model5$y))]
model5.inv=lm((Weight.gained..grams.)^-1~Calories.Consumed)
plot(model5.inv)
summary(model5.inv)
summary(model1)